### ISSUE TITLE
Found endless loop condition in phase 3

### ISSUE THREAD

**Issue by FilixDark** [Chia-Network/chiapos#278]
I have been trying to improve the code a bit for rotational drives, played a bit with the buffer sizes and I am not sure if my changes are causing it atm, or if I just hit the right settings to trigger a bug. Sadly I dont totally understand the whole algorithm yet and I am not a c++ developer, so its utterly painful for me, to read through the code.

But I noticed the same behaviour with the unchanged code as well, sometimes, that in phase 3 the process goes up to 100% on a single core, no further IO and keeps just spinning. 

What happens is, that it reads an entry pos, that doesnt equals current pos,
the value gets cached and current pos gets increased till both match, but in this case, the entry_pos is already smaller than the current_pos. 

Why it happens, I cannot say, but this case is not handled, not throwing any error and causing this endless loop.

first line 261 gets triggert, till current pos becomes bigger than greatest pos  which gets checked in line 235 and totally stops the whole reading process. 

I have right now a version that consistently creates this case and I will try to see, how this value gets written there, but as I said, I don't fully understand yet, how everything is working, so if anyone would like to get in contact and help, that would be totally appreciated.

**Comment by FilixDark** [comment]
Ok, so I found why i constnaly caused this issue in my code. I added async reading to the buffered disk, so it would already preload the next buffer, while its processing the old data. But the special case in phase 2 table 7, where it reads and writes to the same file, caused seeks to execute, while it was reading or writing.  So data got written into the wrong position. At least this is my guess. I added a muted to the buffered disk and locked the internal read and writes and it went through without any issues. 

So maybe there is nothing wrong with the original code and it the hanging processes might still be hardware related, but I think the condition described, is still what keeps the process stuck in an endless loop. wort to put at least a check and throw an exception, that data was inconsistent. It cannot hurt and if  I am right people would at least get some feedback, that something is wrong. 

I had an issue with my memory before on unchanged code. It was set to vendor spec speed  3200, but got my processes stuck, then i underclocked it to 2666 and had no issues.  

If I will find more infos, I will let you know.

**Comment by dosex** [comment]
Interesting. I have observed the same issue on my two plotters as well. Both AMD ryzen machines with 3600 MHz RAM Speed running Linux. One has cheap RAM and on this machine the plotting process got stuck in phase 3 with one core at 100% usage frequently (about 1 in 4 plots). I suspected a RAM issue and clocked it down to 3200 MHz which seamed to have worked, but it still happens from time to time (once every 2 days).

The other machine with the expensive RAM works fine at 3600 MHz most of the time, but it has happened twice to me here as well (done about 600 plots with this one so far).         

I have switched to a custom version of chiapos (https://github.com/pechy/chiapos/tree/combined) about 2 weeks ago which I merged with 1.0.2. The observed behaviour of processes getting stuck in phase 3 happened with the original code and this modified version likewise, so I'd still say it is a RAM issue.
